ianoc/bazelfe
# Use a background process to boost functionality

This is a WIP to boost functionality.

Some new stuff:
-> A detailed robust bazel command line parser
     This is needed to do other things mostly, we need to be able to accurately identify args for commands to bazel vs options, which means we need actually to correctly identify options to run the parsing rules.

-> Background daemon that will do file change tracking
    This lets us have a bazel command trigger on the event of changes in the file system. (basic version is something like ibazel).

-> Test interception that is ‘smart’ and suggests targets
    If no arg is given to test it will offer some test targets to test based on recent files edited.


-> Autotest
   This will make the system run interactively in the console, here when a change occurs we start doing an outward expansion of targets to test based on the change. The order/range of targets here to do is a bit of a WIP. But the general gist is to do something like given a set of dependencies like:

```
MyLibrary.java  -> MyLibraryTest.java
                           |
                           |—-> MyApplicationA —-> MyApplicationATest
                           |—-> MyApplicationB —-> MyApplicationBTest
```
That is we have a library `MyLibrary`, which has a test, and then we have 2 applications that use that library.

We will aim to run this as:
1) Build MyLibrary
2) Build MyLibraryTest
3) Test MyLibraryTest
4) Build MyApplicationA, MyApplicationB
5) Build MyApplicationATest, MyApplicationBTest
6) Test MyApplicationATest, MyApplicationBTest


That is we are going to aim to have the fastest turn around time we can by not trying to build the applications or do anything with them if the tests are broken locally. Then we will expand out to see if building more things works. Finally presuming those tests are expensive we run the tests last.

The problems we are aiming to avoid:
1) No idea what to build based on file edited
2) No idea what the tests are/where they are
3) Not running far more builds/tests than needed `//...`


# Status

Currently the major functionalities listed above have been built.

TODO:
1) Get some beta testing in bigger repos
2) Test myself in something like the bazel repo
3) Add a terminal UI for autotest to make it nicer to look at and be less noisy.
